### PR TITLE
Add ValueToIndexVoxels to simplify using indices instead of masks

### PR DESCRIPTION
**Pull request by mgeplf**

### CODE REVIEW

File: voxcell/voxel_data.py [hunk 1]
```diff
@@ -409,6 +409,43 @@ def __init__(self, *args, **kwargs):
...
+    """Class for efficient access to indices of unique values of the values array.
+
+    Example:
+        fativ = FastApplyToIndexVoxels(br.raw)
+        values_and_funcs= [(i, np.sum if i % 2 else np.mean) for i in fativ.values[:10]]
+        list(fativ.apply(values_and_funcs, density.raw))
+    """
+
+    def __init__(self, values):
+        values = values.ravel()
```
Comment (eleftherioszisis) [inline voxcell/voxel_data.py:422]: How about taking into account the order of the matrix:
```python
def __init__(self, matrix):
    self._order = "C" if matrix.flags.C_CONTIGUOUS else "F"
    values = values.ravel(order=self._order)
```
and then in apply:
```python
    def apply(self, values_and_funcs, voxel_data):
        flat_data = voxel_data.ravel(order=self._order)
        for value, func in values_and_funcs:
            idx = self._get_group_indices_by_value(value)
            yield func(flat_data[idx])
```
If both matrices have the same order, then we don't have to copy when raveling.

File: voxcell/voxel_data.py [hunk 2]
```diff
@@ -409,6 +410,96 @@ def __init__(self, *args, **kwargs):
...
+        self._indices = np.argsort(values, kind="stable")
+        self._offsets = offsets
+        self._mapping = {v: i for i, v in enumerate(uniques)}
+
+    @property
+    def values(self):
+        """List of values that are found in the original volume."""
+        return list(self._mapping)
+
+    def value_to_1d_indices(self, value):
```
Comment (mgeplf) [inline voxcell/voxel_data.py:450]: I made `value_to_1d_indices` and `ravel` public, so people can write their own custom code if they want to

File: voxcell/voxel_data.py [hunk 3]
```diff
@@ -499,6 +505,29 @@ def assign(self, index_voxel_values, voxel_data, inplace=False):
 
         return flat_data.reshape(original_shape, order=self._order)
 
+    def indexed_sum(self, voxel_data):
+        """Calculate the sum of voxel_data values using `values` as a index."""
+        index_values = self.values
+        voxel_values = self.ravel(voxel_data)
+
+        sums = np.zeros(index_values.size, dtype=float)
```
Comment (mgeplf) [inline voxcell/voxel_data.py]: `np.zeros_like(self.values, dtype=float)`?

File: voxcell/voxel_data.py [hunk 4]
```diff
@@ -409,6 +410,133 @@ def __init__(self, *args, **kwargs):
...
+        return sums
+
+    def indexed_count(self):
+        """Calculate the voxel count of each value in `values`."""
+        counts = np.zeros(self.index_size, dtype=int)
+
+        # add.at allows accumulating when the same index is encountered
+        np.add.at(counts, self.codes, 1)
+
+        return counts
```
Comment (eleftherioszisis) [inline voxcell/voxel_data.py:537]: We can alternatively have a second class with these two vectorized methods and an `__init__` of:
```python
values, codes = pd.factorize(values)
```
I have to admit that the don't perfectly fit in this class as we need to store codes just for these two optional methods.

File: voxcell/voxel_data.py (hunk 3 above)
Comment (eleftherioszisis) [reply voxcell/voxel_data.py]: That's outdated code. I don't use the index_values as they are not needed and only store the index size that I use here.

File: voxcell/voxel_data.py [hunk 5]
```diff
@@ -409,6 +409,73 @@ def __init__(self, *args, **kwargs):
...
+        return np.fromiter(self._mapping, dtype=self.index_dtype)
+
+    def value_to_1d_indices(self, value):
+        """Return the indices array indices corresponding to the 'value'.
+
+        Note: These are 1D indices, so the assumption is they are applied to a volume
+        who has been ValueToIndexVoxels::ravel(volume)
+        """
+        if value not in self._mapping:
+            return np.array([], dtype=np.uint64)
```
Comment (eleftherioszisis) [inline voxcell/voxel_data.py:469]:
```suggestion
            return np.array([], dtype=self._indices.dtype)
```

**Review by eleftherioszisis — APPROVED** [review]

**Pull request merged**